### PR TITLE
Fix OAuth token refresh

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

The PR https://github.com/mattermost/mattermost-plugin-zoom/pull/242 makes it so we refresh a user's token whenever we are about to use the token for some operation. The PR's implementation causes an issue when the user is first connecting. When we attempt to fetch an existing token, it fails because there is not one in the kv store. The operation then bails instead of handling the case where the user hasn't connected yet.

This PR makes it so we avoid refreshing the token if this is the first time connecting.

The issue that https://github.com/mattermost/mattermost-plugin-zoom/pull/242 fixes should be tested here as well. I'm concerned that this error was not caught while testing that PR. I'm thinking maybe the plugin may not have been configured properly when testing. Note that these config settings need to be set in order to properly test these two PRs:

- "Enable OAuth" should be set to `true`
- "OAuth by Account Level App (Beta)" should be set to `false`

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-zoom/issues/252

### CODE REVIEW

**Comment by mickmister** [comment]
Note that it's best to review this with whitespace changes off, as it is just wrapping an existing block with an if statement.

File: server/zoom/oauth.go
Comment (catalintomai) [inline server/zoom/oauth.go:155]: @mickmister - is there a way to detect this particular condition "it is likely that this is the **first the time** the user has connected" ? - like if that condition does not apply and there is an error, what can happen? Maybe open another ticket for later to follow-through on this and further refine the handling for this exact condition only?

Comment (mickmister) [reply server/zoom/oauth.go]: @catalintomai Thanks for pointing this out. Initially, I had modified the method `fetchOAuthUserInfo`, which is a more low-level method that fetches the token from the kv store, to signal to its caller when the token doesn't exist. At the moment, it returns an error both when an actual error occurs, and when the token just doesn't exist. I modified this method to return `nil, nil` if the token doesn't exist, then it is up to the caller to differentiate between the two situations (is `err` nil? if not is `token` nil?)

This logic would need to be applied to 5 places, all doing duplicate checking, and potentially causing a nil pointer panic in the future in a new spot that calls this method. Not to mention none of these particular callers need to care if it was an actual error, or the token is simply notexistent, besides this one case where we need to know that.

I was thinking explicitly passing in a boolean to signal "this is a first connect operation" would be cumbersome, or overly explicit. After reading your comment and thinking about it a bit more, I tried it out and it makes the code very readable and intentional. Not much code needs to change either.

I've updated the PR with this strategy, and am satisfied with the solution. Please take another look when you have the chance.

File: server/http.go
Comment (catalintomai) [inline server/http.go:138]: @mickmister 0/5 but it is usually a good practice to add a comment - smth. like `.GetUser(user, true /* first connect */ - using here the C/C# style of comments)` and `.GetUser(user, false /* not the first connect */)` to provide more context to the occasional code reader.

File: server/zoom/oauth.go
Comment (catalintomai) [reply server/zoom/oauth.go]: @mickmister , thanks for following through, liking this much better! Prob. can update the comment as well.

File: server/http.go
Comment (mickmister) [reply server/http.go:138]: @catalintomai I prefer to name the variable in this case, what do you think?

```suggestion
	firstConnect := true
	zoomUser, authErr := client.GetUser(user, firstConnect)
```

Comment (mickmister) [reply server/http.go:138]: Or `firstTimeConnecting` to be more explicit

**Review by catalintomai — APPROVED** [review]

File: server/http.go
Comment (catalintomai) [reply server/http.go]: @mickmister, it is your call - it is a bit of a tradeoff between verbosity and laconism, have to check if golang allows you to have inline assignment e.g. smth like .`GetUser(user, firstConnect = true)`. I've definitely seen version I've mentioned above used countless times though in C# (not sure about golang).

Comment (mickmister) [reply server/http.go]: I've now gone with the variable name strategy I proposed above

**Review by larkox — APPROVED** [review]
LGTM. Any possibility something like this has to be handled also on account level?

**Comment by DHaussermann** [comment]
@mickmister I'm still seeing an issue with this PR deployed that does not seem to affect the production release.
The connection links opens in a new tab but, the authentication flow does not finish. 

Not sure if this is directly related to the issue you're trying to solve but if I'm not mistaken you were able to reproduce it locally.

**Comment by mickmister** [comment]
@DHaussermann I've addressed the issues you pointed out in the most recent commits. Two issues that were fixed:

- When attempting to refresh the user's token, we weren't decrypting the oauth token when fetching from the kv store.
- The kv store contains two entries for each connected user. One is indexed using the MM user ID, the other indexed using the zoom user ID. We were attempting to fetch (and update) the entry indexed for the zoom user ID, but using the MM user ID as the passed in argument.

**Comment by DHaussermann** [comment]
@mickmister The issue seems partially resolved. However the behavior is still inconsistent. On the first attempt users can connect and create meetings. Once they disconnect and need to reconnect I am seeing issues. 
- In 1 case the redirect just hangs in a new browser tab. 
- In another case on a different server, the user see's an error that seems to be coming from Zoom
`{"status":false,"errorCode":4700,"errorMessage":"Invalid redirect: https://dkh-master-ha-14.test.mattermost.cloud/plugins/zoom/oauth2/complete","result":null}`
image I have the 2nd case still in place on a test server if observing this would be useful.

**Comment by mickmister** [comment]
@DHaussermann Please use `pre-release-1.6.0_with-michael-fix` to test this PR.

It is the `1.5.1` release plus https://github.com/mattermost/mattermost-plugin-zoom/pull/242, plus the current state of this PR.

**Comment by DHaussermann** [comment]
@mickmister just to be explicit with were we are on this - I believe next steps are on me. I have a debug build you provided. I will try again to reproduce and collect info when I have a chance.

**Comment by mickmister** [comment]
@levb @DHaussermann This PR is currently blocking the release of master of the zoom plugin. High impact due to refreshing OAuth token

**Comment by DHaussermann** [comment]
@mickmister I am running commit and the issue where the user must reconnect is still occurring.
image

Also I'm still intermittently seeing this issue when I click the link. 
image
Running connect again to generate another connection link then works normally. 
 
 Should we view this as a separate problem or are they intertwined?

**Comment by mickmister** [comment]
> @mickmister I am running commit `640ac12` and the issue where the user must reconnect is still occurring.

@DHaussermann I believe this is fixed with the latest commit now. There was an issue with encrypting the access key before storing, which was affecting the caller of the function that intends to use the unencrypted token.

> Also I'm still intermittently seeing this issue when I click the link. image
> 
> Running connect again to generate another connection link then works normally.
> 
> Should we view this as a separate problem or are they intertwined?

I've seen this but I'm not sure what the reproducible steps are

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Tested multiple time connecting a user connecting and remaining connected for several hours without issue.
- User connection state always seems correct in the webapp
- Tested connect and disconnect several times and the plugin now behaves as expected
- oAuth errors when using connection link did not reoccur on latest commit
LGTM!

Huge thanks @mickmister As of now, I see no issues left to address. Please merge.

**Pull request merged**
